### ISSUE TITLE
Can't create separate S3 bucket storage for individual user

### ISSUE THREAD

**Issue by Harshit-9398** [opendatahub-io/odh-manifests#247]
Hello,
I have been working on a project in which I have been leveraging ODH. My requirement is to have an S3 bucket created for each individual user who logs into JupyterHub. Currently, with the availability of Ceph Nano in the latest release, it is possible to have a single S3 storage bucket for each instance of the OpenDataHub operator, due to which the same S3 bucket has to be accessed by all the users accessing the same JupyterHub instance.
So, is there any way to achieve this functionality in the current scenario of Ceph Nano?
@vpavlin @LaVLaS @crobby @goern

**Comment by guimou** [comment]
Hi!
Would you want to have something like this: https://youtu.be/BE9TvaIpiZk?t=683
TL;DR: It automatically creates a bucket for a user it does not already exist, plus shows it as a folder inside Jupyter.
Code (and explanation) is here: https://github.com/guimou/dataprep-code/tree/master/odh-ocs

That's quite old and was using Object Bucket Claims from Noobaa in this implementation/demo, but can be easily adapted for Ceph Nano. In this case it would just use Object Bucket Claims against the RGW. Plus of course you're not forced to "mount" the bucket inside the notebook, which requires a custom notebook image, you can just push the credentials as env vars. In this case no special modification of the code is necessary, only a custom overlay in ODH deployment to push the config map that the spawner will read (like I'm doing here: https://github.com/guimou/odh-manifests/blob/hmf/jupyterhub/jupyterhub/overlays/shared-pvc/jupyterhub-configmap.yaml).

**Comment by Harshit-9398** [comment]
Hey!
I know this issue has been closed, but I am facing another issue related to this conversation, in which I believe you might be able to help me out.
I am using Rook-Ceph for the purpose of S3 buckets instead of Ceph Nano, and I am able to mount the bucket of a user in their JupyterHub pod. For this, I got the reference from [this](https://github.com/guimou/jupyter-notebooks/blob/s3/minimal-notebook/jupyter_notebook_config.py) file, and I added the corresponding changes in [thoth-station's](https://github.com/thoth-station/s2i-minimal-notebook/blob/master/jupyter_notebook_config.py) file.
I used the below build-config file to create an s2i-minimal-notebook image with support for jupyterlab as well as mounting of S3 buckets for the user:

[build-config.txt](https://github.com/opendatahub-io/odh-manifests/files/5795692/build-config.txt)

I also installed **JupyterLab-git** extension in the resulting image using a DockerFile, and when I spawned this image I faced the below errors:
1) while initializing a git repo using the JupyterHub GUI:
![git init error](https://user-images.githubusercontent.com/36565937/104183077-6a66f980-5437-11eb-98d9-6ea654576211.png)


2) while cloning a git repo using the JupyterHub GUI:
![clone error](https://user-images.githubusercontent.com/36565937/104183025-57542980-5437-11eb-97ab-f6d4b4e33839.png)


But, I am able to do the same operations using Git CLI in the pod's terminal. So, can you please help to find the issue?
@vpavlin @guimou

**Comment by guimou** [comment]
Hey @Harshit-9398 !
Where are you trying to initialize the git repo? Because I don't know if it's related, but there are some caveats with this way of accessing S3 buckets:
- The S3 bucket is not really "mounted" inside the Jupyter environment. It is displayed in the file browser, and you can use it to upload or download files. But that's the Jupyter extension that is taking care of converting those operations to S3 ones.
- So as it is not a mount, it does not exist in the terminal or pod space, it just won't display in a simple "ls" command. It won't also work in a Python notebook with a simple `open` and `f.read`command.
- So if you are trying to use the git extension against such a "folder" I'm pretty sure it will use standard CLI commands, which won't have any effect as the folder does not exist from this point of view.
I know that it can confuse users, that's why some organizations I've worked with chose not to do it in the end (display buckets in the standard tree).
Now if you were trying to do your git operations against another folder, a standard one mounted into the pod, there must be another cause to your issue.

**Comment by Harshit-9398** [comment]
Hey @guimou!
I completely agree with your point that we would not be able to use git in the S3 bucket mounted in the environment.
When I was trying to use git in the `/opt/app-root/src` folder prior to integrating and mounting the S3 bucket as a folder, I was able to perform git operations successfully both by GUI (using the jupyterlab-git extension) and by GIT CLI in the terminal. But after integrating the S3 bucket, I am unable to perform the same operations by GUI and facing the errors whose screenshots I have shared above.
As far as I was able to figure out in this issue, when I am trying to initialize a git repository in the `/opt/app-root/src` folder, it is trying to initialize the repo in the `/` folder due to which it shows permission error.

**Comment by Harshit-9398** [comment]
Hey @guimou,
I am facing another issue (after mounting the S3 buckets in the user pod) with the **Elyra extension of JupyterLab**. The issue seems to be similar to the issue I am facing with the Git extension as I shared above.
I was trying to execute a sample pipeline provided in [Elyra examples repo](https://github.com/elyra-ai/examples/tree/master/pipelines/hello_world) in my environment. But, I am facing the below error when I tried to run it in the JupyterLab notebook images having S3 buckets mounted in them:

![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/36565937/106552504-7e38e380-653d-11eb-97d3-b4e70873dcfe.png)

But, I am able to run the pipeline successfully on JupyterLab notebook images which do not have S3 buckets mounted in them.
As far as I was able to figure out with this issue as well, when I am trying to run the pipeline in images with S3 mounted, the extension is trying to find files in the `/examples/pipelines/` folder which gives the error as shown in the above screenshot, and when I run the pipeline in images without mounting the S3 buckets, it finds the files in the `/opt/app-root/src/examples/pipelines/` folder.  
<br>
Do you have any idea, why is it happening that after mounting the S3 bucket in the notebook image, some extensions are using `/` as their default path rather than `/opt/app-root/src/`?

**Comment by guimou** [comment]
Hi @Harshit-9398 !
Tbh I really don't see why this happens. There must be some PATH variable change for sure that breaks the applications.
I'm sorry but I don't have the environment right now to test. Did you try to spawn a "normal" environment, extract all the env variables, then do the same with an "S3-enabled" environment to compare the output? It may help us figure this out?
Also, have you tried to create an issue here: https://github.com/danielfrg/s3contents ? People there might have stumbled upon the same problem as it's the package used here.

**Comment by Harshit-9398** [comment]
Hi @guimou!
As you suggested, I have checked the environment variables in both cases and I am unable to identify any difference. So, I could not get any clue as to why this is happening.  
I also tried to use the **LargeFileManager** instead of **FileContentsManager** as given in the [**s3contents repo**](https://github.com/danielfrg/s3contents#access-local-files), but that doesn't seem to work as well.

**Comment by Harshit-9398** [comment]
Hi @guimou,  
I raised an issue as suggested by you on the [**s3contents**](https://github.com/danielfrg/s3contents/issues/115) repo, but I haven't got anything from their side till now.  

But, I have another issue when I am trying to upload a file having size > 1 MB in the JupyterLab where S3 buckets are mounted, the files are not getting uploaded properly and their size gets **reduced to anything < 1MB** irrespective of how large their original size is.  

When I tried this in images, where **S3 buckets were not mounted**, the files were uploaded successfully and were of their original size.
